Source/WebCore/platform/graphics/coretext/FontCoreText.cpp

```diff
@@ -530,6 +537,62 @@ RefPtr<Font> Font::platformCreateScaledFont(const FontDescription&, float scaleF
 }
 
+bool supportsOpenTypeFeature(CTFontRef font, CFStringRef featureTag)
+{
+    RetainPtr<CFArrayRef> features = adoptCF(CTFontCopyFeatures(font));
+    CFIndex featureCount = CFArrayGetCount(features.get());
+    for (CFIndex featureIndex = 0; featureIndex < featureCount; ++featureIndex) {
+        auto feature = static_cast<CFDictionaryRef>(CFArrayGetValueAtIndex(features.get(), featureIndex));
+        auto featureTypeIdentifierKey = static_cast<CFNumberRef>(CFDictionaryGetValue(feature, kCTFontFeatureTypeIdentifierKey));
+        if (!featureTypeIdentifierKey)
+            continue;
+
+        int featureTypeIdentifier;
+        CFNumberGetValue(featureTypeIdentifierKey, kCFNumberIntType, &featureTypeIdentifier);
+        if (featureTypeIdentifier != kTextSpacingType)
+            continue;
+
+        auto featureSelectors = static_cast<CFArrayRef>(CFDictionaryGetValue(feature, kCTFontFeatureTypeSelectorsKey));
+        if (!featureSelectors)
+            continue;
+        auto selectorsCount = CFArrayGetCount(featureSelectors);
+        for (CFIndex selectorIndex = 0; selectorIndex < selectorsCount; ++selectorIndex) {
+            auto featureSelector = static_cast<CFDictionaryRef>(CFArrayGetValueAtIndex(featureSelectors, selectorIndex));
+            auto openTypeTag = static_cast<CFStringRef>(CFDictionaryGetValue(featureSelector, kCTFontOpenTypeFeatureTag));
+        if (!openTypeTag)
+            continue;
+        if (CFStringCompare(openTypeTag, featureTag, 0))
+            return true;
+        }
+    }
+    return false;
+}
+
+RefPtr<Font> Font::platformCreateHalfWidthFont() const
+{
+    if (!supportsOpenTypeAlternateHalfWidths())
+        return nullptr;
+
+    RetainPtr<CTFontDescriptorRef> fontDescriptor = adoptCF(CTFontCopyFontDescriptor(getCTFont()));
+    auto size = m_platformData.size();
+    auto fontTraits = CTFontGetSymbolicTraits(getCTFont());
+    int enableHaltValue = 1;
+    auto featureName = CFSTR("halt");
+    auto featureValue = adoptCF(CFNumberCreate(kCFAllocatorDefault, kCFNumberIntType, &enableHaltValue));
+
+    const void* featureValues[] = { featureName, featureValue.get() };
+    auto fontFeatureSettings = adoptCF(CFArrayCreate(kCFAllocatorDefault, featureValues, std::size(featureValues), &kCFTypeArrayCallBacks));
+
+    CFTypeRef fontDescriptorKeys[] = { kCTFontFeatureSettingsAttribute };
+    CFTypeRef fontDescriptorValues[] = { fontFeatureSettings.get() };
+    auto attributes = adoptCF(CFDictionaryCreate(kCFAllocatorDefault, fontDescriptorKeys, fontDescriptorValues, std::size(fontDescriptorValues), &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks));
+
+    auto attributesDescriptor = adoptCF(CTFontDescriptorCreateWithAttributes(attributes.get()));
+    RetainPtr<CTFontRef> halfWidthFont = adoptCF(CTFontCreateCopyWithAttributes(getCTFont(), size, nullptr, attributesDescriptor.get()));
```

Comment:
Why do you explicitly call out a `RetainPtr<>` here, when the preceding line is `auto`?